File: ControleGastosResidenciais.Api/Services/TransacaoService.cs
using ControleGastosResidenciais.Api.Data;
using ControleGastosResidenciais.Api.Models;
using ControleGastosResidenciais.Api.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace ControleGastosResidenciais.Api.Services;

/// <summary>
/// Implementação do serviço de transações
/// </summary>
public class TransacaoService : ITransacaoService
{
    private readonly ControleGastosContext _context;

    public TransacaoService(ControleGastosContext context)
    {
        _context = context;
    }

    public async Task<Transacao> CreateAsync(Transacao transacao)
    {
        // Validar se pessoa existe
        var pessoa = await _context.Pessoas.FindAsync(transacao.PessoaId);
        if (pessoa == null)
            throw new ArgumentException("Pessoa não encontrada");

        // Validar se categoria existe
        var categoria = await _context.Categorias.FindAsync(transacao.CategoriaId);
        if (categoria == null)
            throw new ArgumentException("Categoria não encontrada");

        // Menores de 18 anos só podem fazer despesas
        if (pessoa.Idade < 18 && transacao.Tipo == TipoTransacao.Receita)
            throw new InvalidOperationException("Menores de 18 anos só podem registrar despesas");

        //TODO Validar compatibilidade entre tipo da transação e finalidade da categoria

        _context.Transacoes.Add(transacao);
        await _context.SaveChangesAsync();

        // Reload para incluir as navigation properties
        return await GetByIdAsync(transacao.Id) ?? transacao;
    }

    public async Task<IEnumerable<Transacao>> GetAllAsync()
    {
        return await _context.Transacoes
            .OrderByDescending(t => t.DataCriacao)
            .ToListAsync();
    }

    public async Task<Transacao?> GetByIdAsync(Guid id)
    {
        return await _context.Transacoes.FirstOrDefaultAsync(t => t.Id == id);
    }

    public async Task<IEnumerable<Transacao>> GetByPessoaIdAsync(Guid pessoaId)
    {
        return await _context.Transacoes
            .Where(t => t.PessoaId == pessoaId)
            .OrderByDescending(t => t.DataCriacao)
            .ToListAsync();
    }

}

File: ControleGastosResidenciais.Frontend/src/services/transacaoService.ts
import api from './api';
import { Transacao, CreateTransacaoDto } from '../types';

export const transacaoService = {
  async getAll(): Promise<Transacao[]> {
    const response = await api.get('/transacoes');
    return response.data;
  },

  async getById(id: string): Promise<Transacao> {
    const response = await api.get(`/transacoes/${id}`);
    return response.data;
  },

  async create(transacao: CreateTransacaoDto): Promise<Transacao> {
    const response = await api.post('/transacoes', transacao);
    return response.data;
  },

  async update(id: string, transacao: CreateTransacaoDto): Promise<Transacao> {
    const response = await api.put(`/transacoes/${id}`, transacao);
    return response.data;
  },

  async delete(id: string): Promise<void> {
    await api.delete(`/transacoes/${id}`);
  },

  async getByPessoaId(pessoaId: string): Promise<Transacao[]> {
    const response = await api.get(`/transacoes/pessoa/${pessoaId}`);
    return response.data;
  },

  async getByCategoriaId(categoriaId: string): Promise<Transacao[]> {
    const response = await api.get(`/transacoes/categoria/${categoriaId}`);
    return response.data;
  },
};